tests/id/id.cpp

```diff
@@ -0,0 +1,482 @@
+  CHECK(b * a == idh<D>::get(15, 24, 9));
+  CHECK(a / b == idh<D>::get(1, 2, 1));
+  CHECK(b / a == idh<D>::get(0, 0, 1));
+  CHECK(a % b == idh<D>::get(2, 2, 0));
+  CHECK(b % a == idh<D>::get(3, 3, 0));
+  CHECK(a << b == idh<D>::get(40, 64, 24));
+  CHECK(b << a == idh<D>::get(96, 768, 24));
+  CHECK(a >> b == idh<D>::get(0, 1, 0));
+  CHECK(b >> a == idh<D>::get(0, 0, 0));
+  CHECK((a & b) == idh<D>::get(1, 0, 3));
+  CHECK((b & a) == idh<D>::get(1, 0, 3));
+  CHECK((a | b) == idh<D>::get(7, 11, 3));
+  CHECK((b | a) == idh<D>::get(7, 11, 3));
+  CHECK((a ^ b) == idh<D>::get(6, 11, 0));
+  CHECK((b ^ a) == idh<D>::get(6, 11, 0));
+  CHECK((a && b) == idh<D>::get(1, 1, 1));
+  CHECK((b && a) == idh<D>::get(1, 1, 1));
+  CHECK((a || b) == idh<D>::get(1, 1, 1));
+  CHECK((b || a) == idh<D>::get(1, 1, 1));
+  CHECK((a < b) == idh<D>::get(0, 0, 0));
+  CHECK((b < a) == idh<D>::get(1, 1, 0));
+  CHECK((a > b) == idh<D>::get(1, 1, 0));
+  CHECK((b > a) == idh<D>::get(0, 0, 0));
+  CHECK((a <= b) == idh<D>::get(0, 0, 1));
+  CHECK((b <= a) == idh<D>::get(1, 1, 1));
+  CHECK((a >= b) == idh<D>::get(1, 1, 1));
+  CHECK((b >= a) == idh<D>::get(0, 0, 1));
+
+  CHECK(DEVICE_EVAL(a + b) == idh<D>::get(8, 11, 6));
+  CHECK(DEVICE_EVAL(b + a) == idh<D>::get(8, 11, 6));
+  CHECK(DEVICE_EVAL(a - b) == idh<D>::get(2, 5, 0));
+  CHECK(DEVICE_EVAL(b - a) == idh<D>::get(-2ul, -5ul, 0));
+  CHECK(DEVICE_EVAL(a * b) == idh<D>::get(15, 24, 9));
+  CHECK(DEVICE_EVAL(b * a) == idh<D>::get(15, 24, 9));
+  CHECK(DEVICE_EVAL(a / b) == idh<D>::get(1, 2, 1));
+  CHECK(DEVICE_EVAL(b / a) == idh<D>::get(0, 0, 1));
+  CHECK(DEVICE_EVAL(a % b) == idh<D>::get(2, 2, 0));
+  CHECK(DEVICE_EVAL(b % a) == idh<D>::get(3, 3, 0));
+  CHECK(DEVICE_EVAL(a << b) == idh<D>::get(40, 64, 24));
+  CHECK(DEVICE_EVAL(b << a) == idh<D>::get(96, 768, 24));
+  CHECK(DEVICE_EVAL(a >> b) == idh<D>::get(0, 1, 0));
+  CHECK(DEVICE_EVAL(b >> a) == idh<D>::get(0, 0, 0));
+  CHECK(DEVICE_EVAL(a & b) == idh<D>::get(1, 0, 3));
+  CHECK(DEVICE_EVAL(b & a) == idh<D>::get(1, 0, 3));
+  CHECK(DEVICE_EVAL(a | b) == idh<D>::get(7, 11, 3));
+  CHECK(DEVICE_EVAL(b | a) == idh<D>::get(7, 11, 3));
+  CHECK(DEVICE_EVAL(a ^ b) == idh<D>::get(6, 11, 0));
+  CHECK(DEVICE_EVAL(b ^ a) == idh<D>::get(6, 11, 0));
+  CHECK(DEVICE_EVAL(a && b) == idh<D>::get(1, 1, 1));
+  CHECK(DEVICE_EVAL(b && a) == idh<D>::get(1, 1, 1));
+  CHECK(DEVICE_EVAL(a || b) == idh<D>::get(1, 1, 1));
+  CHECK(DEVICE_EVAL(b || a) == idh<D>::get(1, 1, 1));
+  CHECK(DEVICE_EVAL(a < b) == idh<D>::get(0, 0, 0));
+  CHECK(DEVICE_EVAL(b < a) == idh<D>::get(1, 1, 0));
+  CHECK(DEVICE_EVAL(a > b) == idh<D>::get(1, 1, 0));
+  CHECK(DEVICE_EVAL(b > a) == idh<D>::get(0, 0, 0));
+  CHECK(DEVICE_EVAL(a <= b) == idh<D>::get(0, 0, 1));
+  CHECK(DEVICE_EVAL(b <= a) == idh<D>::get(1, 1, 1));
+  CHECK(DEVICE_EVAL(a >= b) == idh<D>::get(1, 1, 1));
+  CHECK(DEVICE_EVAL(b >= a) == idh<D>::get(0, 0, 1));
+}
+
+#define COMPOUND_OP(operand_value, expr) \
```

Comment:
Why a macro?

Reply:
Same as for `DEVICE_EVAL`, I want to pass an unevaluated `expr` into here.